search/includes/classes/class-versioning.php

```diff
@@ -200,27 +201,68 @@ public function get_inactive_versions( Indexable $indexable ) {
 	 * @return array Array of index versions
 	 */
 	public function get_versions( Indexable $indexable, bool $provide_default = true ) {
-		$versions = get_option( self::INDEX_VERSIONS_OPTION, array() );
+		$versions = [];
+
+		if ( $indexable->global ) {
+			$versions = get_site_option( self::INDEX_VERSIONS_OPTION_GLOBAL, array() );
+		} else {
+			$versions = get_option( self::INDEX_VERSIONS_OPTION, array() );
+		}
 
 		$slug = $indexable->slug;
 
-		if ( ! isset( $versions[ $slug ] ) || ! is_array( $versions[ $slug ] ) || empty( $versions[ $slug ] ) ) {
-			if ( $provide_default ) {
-				return array(
-					1 => array(
-						'number' => 1,
-						'active' => true,
-						'created_time' => null, // We don't know when it was actually created
-						'activated_time' => null,
-					),
+		if ( ! $this->_get_versions_is_slug_defined( $versions, $slug ) ) {
+			// Check deprecated location
+			$deprecated_versions = get_site_option( self::INDEX_VERSIONS_OPTION, array() );
+
+			if ( $this->_get_versions_is_slug_defined( $deprecated_versions, $slug ) ) {
+				// Versions are only stored in the deprecated network storage
+				// TODO remove this deprecated check if it is not executed to simplify the code.
+
+				$message = sprintf(
+					"Application %d - %s found index versions in deprecated global storage for '%s' indexable. I will store versions in correct storage based on global flag.",
+					FILES_CLIENT_SITE_ID,
+					home_url(),
+					$slug
+				);
+
+				\Automattic\VIP\Logstash\log2logstash(
+					array(
+						'severity' => 'warning',
+						'feature' => 'vip_search_versioning',
+						'message' => $message,
+					)
 				);
+
+				$normalized_versions = array_map( array( $this, 'normalize_version' ), $deprecated_versions[ $slug ] );
+
+				// Store versions in correct place to avoid triggering this code again
+				$this->update_versions( $indexable, $normalized_versions );
+
+				return $normalized_versions;
 			} else {
-				return [];
+
+				if ( $provide_default ) {
+					return array(
+						1 => array(
+							'number' => 1,
+							'active' => true,
+							'created_time' => null, // We don't know when it was actually created
+							'activated_time' => null,
+						),
+					);
+				} else {
+					return [];
+				}
 			}
+		} else {
+			// Normalize the versions to ensure consistency (have all fields, etc)
+			return array_map( array( $this, 'normalize_version' ), $versions[ $slug ] );
```

Comment:
In the current implementation, the version is _always_ normalized (it's not behind a conditional) - is this change intentional?

Reply:
Oh I see, this code has been duplicated from another spot...hmm, can we write this so we don't have to duplicate the normalization behavior in multiple places?

Reply:
Hmm I realize now that it's updating the option with the normalized values (so has to be duplicated). In that case, can we extract that functionality to a new function? Can still remove the unnecessary `else` here.

Reply:
My thinking was that the fallback to deprecated solution is temporary so I shouldn't do too much refactoring that wont make sense after that is removed. Where I felt extracting `array_map( array( $this, 'normalize_version' ), $versions[ $slug ] )` belongs. As it seems already kinda ok to read to me. But I don't have strong opinions on that. Although maybe it makes sense to create a function `normalize_versionS` which would handle that mapping over items. What do you think?

With adding the `else`, it just seemed better to be more explicit, that we only ever end up there if the `if` is false. All the branches of that `if` returns directly. Because it is kinda complex if (and will get worse with the network check I guess) it might be easy to overlook though. I am fine with skipping the else if it feels easier on your eyes:)

Reply:
Fair enough, let's leave the normalization stuff. For the `else`, I'd go ahead and remove it since it's redundant right now. If future code makes it more complex, we can add it.

Reply:
Done